bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/checksum/DigestManager.java

```diff
@@ -89,17 +96,48 @@ public static DigestManager instantiate(long ledgerId, byte[] passwd, DigestType
      */
     public ByteBufList computeDigestAndPackageForSending(long entryId, long lastAddConfirmed, long length,
             ByteBuf data) {
-        ByteBuf headersBuffer = PooledByteBufAllocator.DEFAULT.buffer(METADATA_LENGTH + macCodeLength);
-        headersBuffer.writeLong(ledgerId);
-        headersBuffer.writeLong(entryId);
-        headersBuffer.writeLong(lastAddConfirmed);
-        headersBuffer.writeLong(length);
-
-        update(headersBuffer);
-        update(data);
-        populateValueAndReset(headersBuffer);
+        if (this.useV2Protocol) {
+            /*
+             * For V2 protocol, use pooled direct ByteBuf's to avoid object allocation in DigestManager.
+             */
+            ByteBuf headersBuffer = PooledByteBufAllocator.DEFAULT.buffer(METADATA_LENGTH + macCodeLength);
+            headersBuffer.writeLong(ledgerId);
+            headersBuffer.writeLong(entryId);
+            headersBuffer.writeLong(lastAddConfirmed);
+            headersBuffer.writeLong(length);
+
+            update(headersBuffer);
+            update(data);
+            populateValueAndReset(headersBuffer);
+
+            return ByteBufList.get(headersBuffer, data);
+        } else {
+            /*
+             * For V3 protocol, use unpooled heap ByteBuf's (backed by accessible array): The one object
+             * allocation here saves us later allocations when converting to protobuf ByteString.
+             */
+            ByteBuf sendBuffer = Unpooled.buffer(METADATA_LENGTH + macCodeLength + data.readableBytes());
+            sendBuffer.writeLong(ledgerId);
+            sendBuffer.writeLong(entryId);
+            sendBuffer.writeLong(lastAddConfirmed);
+            sendBuffer.writeLong(length);
+
+            update(sendBuffer);
+            update(data);
+            populateValueAndReset(sendBuffer);
+
+            if (data.hasArray()) {
+                // fast path: single copy into existing array
+                sendBuffer.writeBytes(data.array(), data.arrayOffset(), data.readableBytes());
+            } else {
+                // fall-back to slow path with extra object allocation and copy
```

Comment:
the logic here seems to be wrong. I think you need to write data into the buffer, no?

any reason why you can't just call `sendBuffer.writeBytes(data, data.readerIndex(), data.readableBytes())` to replace line 129-137?

https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#writeBytes-io.netty.buffer.ByteBuf-int-int-

Reply:
This "transfers the specified source buffer's data to this buffer". The source is "data", our payload, and the destination is "sendBuffer", the buffer containing header, checksum, and payload. Am I missing something?

But I think we can replace the entire block by
sendBuffer.writeBytes(data);

It will figure out the offset/length itself. The distinction isn't necessary indeed.

Reply:
I haven't yet figured out the Netty API and its use in bookkeeper.

It seems to me that writeBytes(ByteBuf src) will consider the source buffers reader index and readableBytes, and according to API doc "This method is basically same with {@link #writeBytes(ByteBuf, int, int)}, except that this method increases the {@code readerIndex} of the source buffer by the number of the transferred bytes while {@link #writeBytes(ByteBuf, int, int)} does not."

I'm not sure whether this difference matters or not... Are we reading from "data" (they payload) elsewhere again? If yes, I suppose we have to use writeBytes(ByteBuf, int, int) instead. Else #writeBytes(ByteBuf) would be just fine, right?

Reply:
`sendBuffer.writeBytes(data)` is basically `sendBuffer.writeBytes(data, int, int)`. `writeBytes(data)` will modify the readerIndex data. typically the practice is if we don't need to modify the reader/writer index for a bytebuf, try not to modify it. that's much clearer